servers/zms/src/main/java/com/yahoo/athenz/zms/DBService.java

```diff
@@ -6402,6 +6403,147 @@ void enforceGroupUserAuthorityRestrictions(final String domainName, final String
+           }
+        }
+        return principals;
+    }
+
+    /**
+     * This method toggles state for supplied Principals based on the flag in DB
+     * as well as modifies memberships of all roles and groups of current principal(s)
+     * @param changedPrincipals List of Principals from User Authority
+     * @param suspended boolean indicating principal's state
+     */
+    void updatePrincipalByStateFromAuthority(List<Principal> changedPrincipals, boolean suspended) {
+        final String caller = "updatePrincipalByStateFromAuthority";
+        List<Principal> updatedUsers = new ArrayList<>();
+        if (!changedPrincipals.isEmpty()) {
+            try (ObjectStoreConnection con = store.getConnection(true, true)) {
+
+                // first lets update the new state in DB
+                for (Principal changedPrincipal : changedPrincipals) {
+                    updatePrincipalState(suspended, updatedUsers, con, changedPrincipal);
+                }
+                // if new state is updated successfully
+                // then we need to modify all roles and groups where given principal is member of
+                if (!updatedUsers.isEmpty()) {
+                    for (Principal updatedUser : updatedUsers) {
+                        try {
+                            updateRoleMembershipsByPrincipalState(suspended, caller, con, updatedUser);
+                            updateGroupMembershipByPrincipalState(suspended, caller, con, updatedUser);
+                        } catch (ResourceException ex) {
+                            if (ex.getCode() == ResourceException.NOT_FOUND) {
+                                continue;
+                            }
+                            throw ex;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private void updateGroupMembershipByPrincipalState(boolean suspended, String caller, ObjectStoreConnection con, Principal updatedUser) {
+        List<GroupMember> groupMembersWithUpdatedState;
+        GroupMember groupMember;
+        DomainGroupMember domainGroupMember;
+        Set<GroupMember> updatedGroups = new HashSet<>();
+
+        domainGroupMember = con.getPrincipalGroups(updatedUser.getFullName(), null);
+        if (!domainGroupMember.getMemberGroups().isEmpty()) {
+            for (GroupMember currentGroup : domainGroupMember.getMemberGroups()) {
+                groupMember = new GroupMember();
+                groupMember.setMemberName(updatedUser.getFullName());
+                if (suspended) {
+                    groupMember.setSystemDisabled(currentGroup.getSystemDisabled() & ~Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                } else {
+                    groupMember.setSystemDisabled(currentGroup.getSystemDisabled() | Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                }
+                groupMembersWithUpdatedState = new ArrayList<>();
```

Comment:
no need to create an array list with one member all the time. Instead use Collections.singletonList()